shared/src/main/scala/interpreters.scala

```diff
@@ -54,19 +49,83 @@ trait FetchInterpreters {
       def apply[A](fa: FetchOp[A]): FetchInterpreter[M]#f[A] = {
         StateT[M, FetchEnv, A] { env: FetchEnv =>
           fa match {
-            case FetchError(e) => M.raiseError(e)
-            case Cached(a)     => M.pure((env, a))
-            case Concurrent(concurrentRequests) => {
+            case Thrown(e)  => M.raiseError(FetchException(e))
+            case Fetched(a) => M.pure((env, a))
+            case one @ FetchOne(id, ds) => {
                 val startRound = System.nanoTime()
                 val cache      = env.cache
 
-                val requests: List[FetchRequest[Any, Any]] =
-                  pendingRequests(concurrentRequests, cache)
+                cache
+                  .get[A](ds.identity(id))
+                  .fold[M[(FetchEnv, A)]](
+                      M.flatMap(M.runQuery(ds.fetchOne(id)))((res: Option[A]) => {
+                        val endRound = System.nanoTime()
+                        res.fold[M[(FetchEnv, A)]](
+                            M.raiseError(
+                                NotFound(env, one)
+                            )
+                        )(result => {
+                          val endRound = System.nanoTime()
+                          val newCache = cache.update(ds.identity(id), result)
+                          val round    = Round(cache, one, result, startRound, endRound)
+                          M.pure(env.evolve(round, newCache) -> result)
+                        })
+                      })
+                  )(cached => {
+                    val endRound = System.nanoTime()
+                    M.pure(env -> cached)
+                  })
+              }
+            case many @ FetchMany(ids, ds) => {
+                val startRound = System.nanoTime()
+                val cache      = env.cache
+                val newIds     = many.missingIdentities(cache)
+                val result     = ids.unwrap.flatMap(id => cache.get(ds.identity(id)))
+
+                if (newIds.isEmpty)
+                  M.pure(env -> result)
+                else {
+                  M.flatMap(M.runQuery(ds
+                            .asInstanceOf[DataSource[I, A]]
```

Comment:
Is there a way to get rid of these casts? If we are not sure we should add an issue to get it cleaned up a bit.